File: api/src/resource/repository.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use uuid::Uuid;

#[derive(ApiResource, Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct RepositoryResource {
    pub id: Uuid,
    pub name: String,
    pub owner: String,
    pub visibility: String,
    pub created_at: DateTime<Utc>,
}

#[derive(ApiResource, Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct RepositoryCommitsResource {
    pub commits: Vec<RepositoryCommitResource>,
    pub has_next: bool,
}

#[derive(ApiResource, Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct RepositoryCommitResource {
    pub sha: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub parent_sha: Option<String>,
    pub message: String,
    pub date: DateTime<Utc>,
    pub author: CommitAuthorResource,
}

#[derive(ApiResource, Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct CommitAuthorResource {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub id: Option<Uuid>,
    pub name: String,
    pub email: String,
}

#[derive(ApiResource, Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct RepositoryFileResource {
    pub ref_name: String,
    pub path: String,
    pub commit_sha: String,
    pub sha: String,
    pub content: String,
    pub encoding: String,
}

#[derive(ApiResource, Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct RepositoryPreviewResource {
    pub ref_name: String,
    pub commit_sha: String,
    pub entries: Vec<RepositoryPreviewEntryResource>,
}

#[derive(ApiResource, Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct RepositoryPreviewEntryResource {
    pub path: String,
    pub name: String,
    pub sha: String,
    pub preview: Option<FilePreviewResource>,
}

#[derive(ApiResource, Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct FilePreviewResource {
    pub content: String,
    pub total_lines: u32,
    pub preview_lines: u32,
    pub truncated: bool,
    pub encoding: String,
}

#[derive(ApiResource, Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct RepositoryTreeResource {
    pub ref_name: String,
    pub commit_sha: String,
    pub entries: Vec<RepositoryTreeEntryResource>,
}

#[derive(ApiResource, Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct RepositoryTreeEntryResource {
    pub path: String,
    pub name: String,
    pub entry_type: String,
    pub sha: String,
    pub commit: RepositoryCommitResource,
}

#[derive(ApiResource, Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct RepositoryCommitStatResource {
    pub path: String,
    pub lines_added: u32,
    pub lines_removed: u32,
}

#[derive(ApiResource, Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct RepositoryCommitDiffResource {
    pub diff: RepositoryDiffResource,
    pub left: Option<RepositoryFileResource>,
    pub right: Option<RepositoryFileResource>,
}

#[derive(ApiResource, Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct RepositoryDiffResource {
    pub lines_added: u32,
    pub lines_removed: u32,
    pub hunks: Vec<DiffHunkResource>,
}

pub type DiffHunkResource = Vec<DiffPairResource>;

#[derive(ApiResource, PartialEq, Eq, Debug, Clone, Serialize, Deserialize)]
pub struct DiffPairResource {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub lhs: Option<DiffLineResource>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub rhs: Option<DiffLineResource>,
}

#[derive(ApiResource, PartialEq, Eq, Debug, Clone, Serialize, Deserialize)]
pub struct DiffLineResource {
    pub line_number: u32,
    pub changes: Vec<DiffChangeResource>,
}

#[derive(ApiResource, PartialEq, Eq, Debug, Clone, Serialize, Deserialize)]
pub struct DiffChangeResource {
    pub start: u32,
    pub end: u32,
    pub content: String,
    pub highlight: SyntaxHighlight,
}

#[derive(ApiResource, Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct RepositoryPermissionResource {
    pub permission: String,
}

#[derive(ApiResource, PartialEq, Eq, Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum SyntaxHighlight {
    Delimiter,
    Normal,
    String,
    Type,
    Comment,
    Keyword,
    TreeSitterError,
}

File: gitdot-api-ts/src/resource/repository.ts
import { z } from "zod";

export const SyntaxHighlight = z.enum([
  "delimiter",
  "normal",
  "string",
  "type",
  "comment",
  "keyword",
  "tree_sitter_error",
]);
export type SyntaxHighlight = z.infer<typeof SyntaxHighlight>;

export const DiffChangeResource = z.object({
  start: z.number().int(),
  end: z.number().int(),
  content: z.string(),
  highlight: SyntaxHighlight,
});
export type DiffChangeResource = z.infer<typeof DiffChangeResource>;

export const DiffLineResource = z.object({
  line_number: z.number().int(),
  changes: z.array(DiffChangeResource),
});
export type DiffLineResource = z.infer<typeof DiffLineResource>;

export const DiffPairResource = z.object({
  lhs: DiffLineResource.optional(),
  rhs: DiffLineResource.optional(),
});
export type DiffPairResource = z.infer<typeof DiffPairResource>;

export const DiffHunkResource = z.array(DiffPairResource);
export type DiffHunkResource = z.infer<typeof DiffHunkResource>;

export const RepositoryDiffResource = z.object({
  lines_added: z.number().int(),
  lines_removed: z.number().int(),
  hunks: z.array(DiffHunkResource),
});
export type RepositoryDiffResource = z.infer<typeof RepositoryDiffResource>;

export const RepositoryFileResource = z.object({
  ref_name: z.string(),
  path: z.string(),
  commit_sha: z.string(),
  sha: z.string(),
  content: z.string(),
  encoding: z.string(),
});
export type RepositoryFileResource = z.infer<typeof RepositoryFileResource>;

export const RepositoryCommitDiffResource = z.object({
  diff: RepositoryDiffResource,
  left: RepositoryFileResource.nullable(),
  right: RepositoryFileResource.nullable(),
});
export type RepositoryCommitDiffResource = z.infer<
  typeof RepositoryCommitDiffResource
>;

export const CommitAuthorResource = z.object({
  id: z.string().uuid().optional(),
  name: z.string(),
  email: z.string(),
});
export type CommitAuthorResource = z.infer<typeof CommitAuthorResource>;

export const RepositoryCommitResource = z.object({
  sha: z.string(),
  parent_sha: z.string().optional(),
  message: z.string(),
  date: z.string().datetime(),
  author: CommitAuthorResource,
});
export type RepositoryCommitResource = z.infer<typeof RepositoryCommitResource>;

export const RepositoryCommitsResource = z.object({
  commits: z.array(RepositoryCommitResource),
  has_next: z.boolean(),
});
export type RepositoryCommitsResource = z.infer<
  typeof RepositoryCommitsResource
>;

export const FilePreviewResource = z.object({
  content: z.string(),
  total_lines: z.number().int(),
  preview_lines: z.number().int(),
  truncated: z.boolean(),
  encoding: z.string(),
});
export type FilePreviewResource = z.infer<typeof FilePreviewResource>;

export const RepositoryPreviewEntryResource = z.object({
  path: z.string(),
  name: z.string(),
  sha: z.string(),
  preview: FilePreviewResource.nullable(),
});
export type RepositoryPreviewEntryResource = z.infer<
  typeof RepositoryPreviewEntryResource
>;

export const RepositoryPreviewResource = z.object({
  ref_name: z.string(),
  commit_sha: z.string(),
  entries: z.array(RepositoryPreviewEntryResource),
});
export type RepositoryPreviewResource = z.infer<
  typeof RepositoryPreviewResource
>;

export const RepositoryTreeEntryResource = z.object({
  path: z.string(),
  name: z.string(),
  entry_type: z.string(),
  sha: z.string(),
  commit: RepositoryCommitResource,
});
export type RepositoryTreeEntryResource = z.infer<
  typeof RepositoryTreeEntryResource
>;

export const RepositoryTreeResource = z.object({
  ref_name: z.string(),
  commit_sha: z.string(),
  entries: z.array(RepositoryTreeEntryResource),
});
export type RepositoryTreeResource = z.infer<typeof RepositoryTreeResource>;

export const RepositoryCommitStatResource = z.object({
  path: z.string(),
  lines_added: z.number().int(),
  lines_removed: z.number().int(),
});
export type RepositoryCommitStatResource = z.infer<
  typeof RepositoryCommitStatResource
>;

export const RepositoryPermissionResource = z.object({
  permission: z.string(),
});
export type RepositoryPermissionResource = z.infer<
  typeof RepositoryPermissionResource
>;

export const RepositoryResource = z.object({
  id: z.string().uuid(),
  name: z.string(),
  owner: z.string(),
  visibility: z.string(),
  created_at: z.string().datetime(),
});
export type RepositoryResource = z.infer<typeof RepositoryResource>;
